developer_docs/ast_nodes.md

```diff
@@ -0,0 +1,73 @@
+
+### Precision
+
+The precision indicates the precision of the datatype. This number is related to the number of bytes that the datatype takes up in memory (e.g. `float64` has precision = 8 as it takes up 8 bytes, `complex128` has precision = 8 as it is comprised of two `float64` objects). The precision is equivalent to the `kind` parameter in Fortran.
+
+In Python the precision of some types depends on the system where the code is run. This is notably the case for integers which have a precision of 4 on Windows but a precision of 8 on Linux and MacOS. This is the case for native types. In order to differentiate these types from the fixed-precision objects provided by NumPy, the precision -1 is used to denote the default precision.
+
+### Rank
+
+The rank of an array is the number of dimensions in the array. This object is an integer.
+
+### Shape
+
+The shape of an array indicates the number of elements in each dimension of the array. This property is a tuple. The tuple contains `TypedAstNode`s. If the shape is known then these objects will be `LiteralInteger`s. If the shape is unknown then these objects will be `PyccelArrayShapeElement`s.
+
+### Order
+
+The order indicates how an array is laid out in memory. This can either be row-major (C-style) ordering or column-major (Fortran-style) ordering. For more information about this, please see the [dedicated documentation](./order_docs.md).
+
+## Pyccel Internal Function
+
+The class `pyccel.ast.internals.PyccelInternalFunction` is a super class. This class should never be used directly but provides functionalities which are common to certain AST objects. These AST nodes are those which describe functions which are supported by Pyccel. For example it is used for functions from the `math` library, the `cmath` library, the `numpy` library, etc. `PyccelInternalFunction` inherits from `TypedAstNode`. The type information for the sub-class describes the type of the result of the function.
+
+Objects inheriting from `PyccelInternalFunction` are created from a `FunctionCall` object resulting from the syntactic stage. The `PyccelInternalFunction` objects are initialised by passing the contents of the `FunctionCallArgument`s used in the `FunctionCall` to the constructor. It is therefore important that the names of the arguments and keyword arguments of any classes inherited from `PyccelInternalFunction` match the names of the arguments and keyword arguments of the function being described.
```

Comment:
Not sure what you mean with "Objects inheriting from `PyccelInternalFunction`". Did you mean "Instances of the subclasses of `PyccelInternalFunction`".?

Reply:
Yes. Fixed 6a13e18fc